tests/unit/_internal/runner/test_strategy.py

```diff
@@ -17,23 +17,73 @@ class GPURunnable(bentoml.Runnable):
     SUPPORTED_RESOURCES = ("nvidia.com/gpu",)
 
 
+class SingleThreadRunnable(bentoml.Runnable):
+    SUPPORTED_RESOURCES = ("cpu",)
+    SUPPORTS_CPU_MULTI_THREADING = False
+
+
+class MultiThreadRunnable(bentoml.Runnable):
+    SUPPORTED_RESOURCES = ("cpu",)
+    SUPPORTS_CPU_MULTI_THREADING = True
+
+
 def unvalidated_get_resource(x: t.Dict[str, t.Any], y: str):
     return get_resource(x, y, validate=False)
 
 
 def test_default_gpu_strategy(monkeypatch: MonkeyPatch):
     monkeypatch.setattr(strategy, "get_resource", unvalidated_get_resource)
-    assert DefaultStrategy.get_worker_count(GPURunnable, {"nvidia.com/gpu": 2}) == 2
+    assert DefaultStrategy.get_worker_count(GPURunnable, {"nvidia.com/gpu": 2}, 1) == 2
+    assert DefaultStrategy.get_worker_count(GPURunnable, {"nvidia.com/gpu": 2}, 2) == 4
     assert pytest.raises(
-        ValueError, DefaultStrategy.get_worker_count, GPURunnable, {"nvidia.com/gpu": 0}
+        ValueError,
+        DefaultStrategy.get_worker_count,
+        GPURunnable,
+        {"nvidia.com/gpu": 0},
+        1,
     )
     assert (
-        DefaultStrategy.get_worker_count(GPURunnable, {"nvidia.com/gpu": [2, 7]}) == 2
+        DefaultStrategy.get_worker_count(GPURunnable, {"nvidia.com/gpu": [2, 7]}, 1)
+        == 2
+    )
+    assert (
+        DefaultStrategy.get_worker_count(GPURunnable, {"nvidia.com/gpu": [2, 7]}, 2)
+        == 4
     )
 
-    envs = DefaultStrategy.get_worker_env(GPURunnable, {"nvidia.com/gpu": 2}, 0)
+    envs = DefaultStrategy.get_worker_env(GPURunnable, {"nvidia.com/gpu": 2}, 1, 0)
     assert envs.get("CUDA_VISIBLE_DEVICES") == "0"
-    envs = DefaultStrategy.get_worker_env(GPURunnable, {"nvidia.com/gpu": 2}, 1)
+    envs = DefaultStrategy.get_worker_env(GPURunnable, {"nvidia.com/gpu": 2}, 1, 1)
     assert envs.get("CUDA_VISIBLE_DEVICES") == "1"
-    envs = DefaultStrategy.get_worker_env(GPURunnable, {"nvidia.com/gpu": [2, 7]}, 1)
+    envs = DefaultStrategy.get_worker_env(GPURunnable, {"nvidia.com/gpu": [2, 7]}, 1, 1)
     assert envs.get("CUDA_VISIBLE_DEVICES") == "7"
+
+    envs = DefaultStrategy.get_worker_env(GPURunnable, {"nvidia.com/gpu": 2}, 2, 0)
+    assert envs.get("CUDA_VISIBLE_DEVICES") == "0"
+    envs = DefaultStrategy.get_worker_env(GPURunnable, {"nvidia.com/gpu": 2}, 2, 1)
+    assert envs.get("CUDA_VISIBLE_DEVICES") == "0"
+    envs = DefaultStrategy.get_worker_env(GPURunnable, {"nvidia.com/gpu": 2}, 2, 2)
+    assert envs.get("CUDA_VISIBLE_DEVICES") == "1"
+    envs = DefaultStrategy.get_worker_env(GPURunnable, {"nvidia.com/gpu": [2, 7]}, 2, 1)
+    assert envs.get("CUDA_VISIBLE_DEVICES") == "2"
```

Comment:
👍

maybe we can add another case for the last configuration that will return CUDA_VISIBLE_DEVICES==7